Rework runner strategy in `MooseControl`, add testing

loganharbour opened the pull request:
Refs #31755

- Renames the `MooseControl` python module to `moosecontrol`
- Moves initialization and finalization in `MooseControl` to a separate runner class

loganharbour committed:
Begin work on separating the runner strategy in the MooseControl

refs #31755

loganharbour committed:
Rename MooseControl module to moosecontrol

refs #31755

loganharbour committed:
Clean up module paths

refs #31755

loganharbour committed:
Remove main() from tests; can't be used with pytest

refs #31755

loganharbour committed:
Begin work on writing new MooseControl

refs #31755

loganharbour committed:
Fixes for python<3.12

refs #31755

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3311955/), step Versioner verify on 45f6b19 wanted to post the following:

### Versioner templates

Found 14 templates, 0 failed


### Versioner influential files

Found 58 influential files, 3 changed, 0 added, 0 removed

| package   | status   | file                                    |
|-----------|----------|-----------------------------------------|
| tools     | CHANGE   | conda/tools/meta.yaml                   |
| moose-dev | CHANGE   | conda/moose-dev/conda_build_config.yaml |
| moose-dev | CHANGE   | conda/moose-dev/meta.yaml               |

### Versioner versions

Found 9 packages, 2 changed, 0 failed

| package   | status   | hash    | to hash   | version    | to version   |
|-----------|----------|---------|-----------|------------|--------------|
| tools     | CHANGE   | d73012a | 1334a18   | 2025.06.26 | 2025.10.20   |
| moose-dev | CHANGE   | 32e04eb | e2e4cee   | 2025.09.18 | 2025.10.20   |

loganharbour committed:
Satisfy pylint and pylance

refs #31755

loganharbour committed:
Find the executable automatically if we can

refs #31755

loganharbour committed:
Add set_controllable_matrix

refs #31755

loganharbour committed:
Add wait()

refs #31755

loganharbour committed:
Rename method

loganharbour committed:
Add get_return_code()

refs #31755

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3312341/), step Results summary on 8ae42c5 wanted to post the following:

## Framework test summary

### Removed tests
None
### Added tests
None
### Run time changes
| Test                                          |   Base (s) |   Head (s) | +/-     |
|-----------------------------------------------|------------|------------|---------|
| `multiapps/command_line.common_vector_single` |       0.97 |       1.57 | +61.11% |
| `fvics/function_ic.parsed_function`           |       0.88 |       1.36 | +55.07% |
## Modules test summary

### Removed tests
None
### Added tests
None
### Run time changes
| Test                                                                              |   Base (s) |   Head (s) | +/-     |
|-----------------------------------------------------------------------------------|------------|------------|---------|
| `stochastic_tools/test:web_server_control.stochastic_control/batch_restore_multi` |       2.77 |       1.06 | -61.85% |
| `stochastic_tools/test:web_server_control.stochastic_control/batch_restore`       |       3.29 |       1.08 | -67.29% |
| `stochastic_tools/test:web_server_control.stochastic_control/batch_reset`         |       3.67 |       1.05 | -71.38% |
| `stochastic_tools/test:web_server_control.stochastic_control/normal`              |       4.95 |       1.3  | -73.78% |
| `stochastic_tools/test:web_server_control.stochastic_control/batch_reset_multi`   |       4.19 |       1.06 | -74.80% |

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/3312274/), step Docs: sync website on 8ae42c5 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/31756/site)

This comment will be updated on new commits.

loganharbour committed:
Make test happy for python<3.12

refs #31755

loganharbour committed:
Add --no-moose option and guard tests

refs #31755

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3313377/), step Versioner verify on c37d9a6 wanted to post the following:

### Versioner templates

Found 14 templates, 0 failed


### Versioner influential files

Found 58 influential files, 3 changed, 0 added, 0 removed

| package   | status   | file                                    |
|-----------|----------|-----------------------------------------|
| tools     | CHANGE   | conda/tools/meta.yaml                   |
| moose-dev | CHANGE   | conda/moose-dev/conda_build_config.yaml |
| moose-dev | CHANGE   | conda/moose-dev/meta.yaml               |

### Versioner versions

Found 9 packages, 2 changed, 0 failed

| package   | status   | hash    | to hash   | version    | to version   |
|-----------|----------|---------|-----------|------------|--------------|
| tools     | CHANGE   | d73012a | 1334a18   | 2025.06.26 | 2025.10.20   |
| moose-dev | CHANGE   | 32e04eb | e2e4cee   | 2025.09.18 | 2025.10.20   |

loganharbour committed:
Replace MOOSE_EXE env var with --moose-exe

loganharbour committed:
Use the directory from the test

loganharbour committed:
Use a common test base

loganharbour committed:
Cleanup random socket

loganharbour committed:
Shorten the random socket path

loganharbour committed:
Simplify copy

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3313578/), step Results summary on 5322277 wanted to post the following:

## Framework test summary

### Removed tests
None
### Added tests
None
### Run time changes
| Test                                       |   Base (s) |   Head (s) | +/-     |
|--------------------------------------------|------------|------------|---------|
| `predictors/simple.test`                   |       0.85 |       1.39 | +63.32% |
| `tag.not_zeroed_vector_tag`                |       0.71 |       1.07 | +52.18% |
| `multiapps/sub_cycling.test_negative_time` |       1.01 |       1.53 | +50.66% |
## Modules test summary

### Removed tests
None
### Added tests
None
### Run time changes
| Test                                                                              |   Base (s) |   Head (s) | +/-     |
|-----------------------------------------------------------------------------------|------------|------------|---------|
| `solid_mechanics/test:drucker_prager.small2_outer_tip`                            |       1.71 |       2.67 | +56.18% |
| `stochastic_tools/test:web_server_control.stochastic_control/batch_restore_multi` |       2.77 |       1.05 | -62.24% |
| `stochastic_tools/test:web_server_control.stochastic_control/batch_restore`       |       3.29 |       1.07 | -67.46% |
| `stochastic_tools/test:web_server_control.stochastic_control/batch_reset`         |       3.67 |       1.1  | -70.01% |
| `stochastic_tools/test:web_server_control.stochastic_control/normal`              |       4.95 |       1.3  | -73.76% |
| `stochastic_tools/test:web_server_control.stochastic_control/batch_reset_multi`   |       4.19 |       1.04 | -75.15% |

loganharbour committed:
Setup logging in MooseControl

refs #31755

loganharbour committed:
Use float type

loganharbour committed:
Wait for the control in initialize()

loganharbour committed:
Use new MooseControl in tests

refs #31755

loganharbour committed:
Remove old MooseControl

refs #31755

loganharbour committed:
Set docstring

loganharbour committed:
Rename file

loganharbour committed:
Remove unused import

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3314629/), step Versioner verify on 23103a2 wanted to post the following:

### Versioner templates

Found 14 templates, 0 failed


### Versioner influential files

Found 58 influential files, 3 changed, 0 added, 0 removed

| package   | status   | file                                    |
|-----------|----------|-----------------------------------------|
| tools     | CHANGE   | conda/tools/meta.yaml                   |
| moose-dev | CHANGE   | conda/moose-dev/conda_build_config.yaml |
| moose-dev | CHANGE   | conda/moose-dev/meta.yaml               |

### Versioner versions

Found 9 packages, 2 changed, 0 failed

| package   | status   | hash    | to hash   | version    | to version   |
|-----------|----------|---------|-----------|------------|--------------|
| tools     | CHANGE   | d73012a | 1334a18   | 2025.06.26 | 2025.10.20   |
| moose-dev | CHANGE   | 32e04eb | e2e4cee   | 2025.09.18 | 2025.10.20   |

loganharbour committed:
Formalize the context manager

refs #31755

loganharbour committed:
Cleanup test: remove unused vars, formalize timeouts

refs #31755

loganharbour committed:
Don't allow initialize() more than once

refs #31755

loganharbour committed:
Don't use shred default

refs #31755

loganharbour committed:
Improve docstring

loganharbour committed:
Add initialize property to MooseControl

loganharbour committed:
Add live test for MooseControl

refs #31755

loganharbour committed:
Wait for control to be waiting

refs #31755

loganharbour committed:
Use more MagicMock

refs #31755

loganharbour committed:
Tighten the timeouts

refs #31755

loganharbour committed:
Remove extra space

loganharbour committed:
Refactor StochasticControl to work with new MooseControl

refs #31755

loganharbour committed:
Satisfy pylint

refs #31755

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3316173/), step Versioner verify on d5160aa wanted to post the following:

### Versioner templates

Found 14 templates, 0 failed


### Versioner influential files

Found 58 influential files, 3 changed, 0 added, 0 removed

| package   | status   | file                                    |
|-----------|----------|-----------------------------------------|
| tools     | CHANGE   | conda/tools/meta.yaml                   |
| moose-dev | CHANGE   | conda/moose-dev/conda_build_config.yaml |
| moose-dev | CHANGE   | conda/moose-dev/meta.yaml               |

### Versioner versions

Found 9 packages, 2 changed, 0 failed

| package   | status   | hash    | to hash   | version    | to version   |
|-----------|----------|---------|-----------|------------|--------------|
| tools     | CHANGE   | d73012a | 1334a18   | 2025.06.26 | 2025.10.20   |
| moose-dev | CHANGE   | 32e04eb | e2e4cee   | 2025.09.18 | 2025.10.20   |

loganharbour committed:
Remove pylint and pylance notations

loganharbour committed:
Update headers to satisfy black

loganharbour committed:
Apply black format

loganharbour committed:
Satisfy ruff

loganharbour committed:
Report timing

loganharbour committed:
Clean up get_pid() logic and speed up test

loganharbour committed:
Speed up tests

loganharbour committed:
Check for no warnings by default

loganharbour committed:
Make test faster

loganharbour committed:
Make test faster

loganharbour committed:
Simplify test

loganharbour committed:
Simplify test

loganharbour committed:
Just look for at least one

loganharbour committed:
Simplify skip message

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3317668/), step Versioner verify on 35c8600 wanted to post the following:

### Versioner templates

Found 14 templates, 0 failed


### Versioner influential files

Found 58 influential files, 3 changed, 0 added, 0 removed

| package   | status   | file                                    |
|-----------|----------|-----------------------------------------|
| tools     | CHANGE   | conda/tools/meta.yaml                   |
| moose-dev | CHANGE   | conda/moose-dev/conda_build_config.yaml |
| moose-dev | CHANGE   | conda/moose-dev/meta.yaml               |

### Versioner versions

Found 9 packages, 2 changed, 0 failed

| package   | status   | hash    | to hash   | version    | to version   |
|-----------|----------|---------|-----------|------------|--------------|
| tools     | CHANGE   | d73012a | 1334a18   | 2025.06.26 | 2025.10.20   |
| moose-dev | CHANGE   | 32e04eb | e2e4cee   | 2025.09.18 | 2025.10.20   |

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3317730/), step Results summary on 35c8600 wanted to post the following:

## Framework test summary

### Removed tests
None
### Added tests
None
### Run time changes
| Test                                                                                 |   Base (s) |   Head (s) | +/-     |
|--------------------------------------------------------------------------------------|------------|------------|---------|
| `meshgenerators/sphere_mesh_generator.hex27`                                         |       0.86 |       1.39 | +61.54% |
| `materials/functor_conversion.vector`                                                |       0.89 |       1.43 | +60.81% |
| `time_steppers/timesequence_stepper.general_restart/steady_for_restart`              |       0.77 |       1.23 | +58.65% |
| `transfers/general_field/nearest_node/boundary.3d_boxes/projection_needed_receiving` |       1.3  |       2    | +53.54% |
| `kernels/coupled_time_derivative.adtestdirichlet`                                    |       0.88 |       1.36 | +53.51% |
| `functions/piecewise_constant.piecewise_constant_from_json`                          |       0.87 |       1.32 | +51.59% |
## Modules test summary

### Removed tests
None
### Added tests
None
### Run time changes
None

loganharbour committed:
Add testing for cleanup() live

refs #31755

loganharbour committed:
Add test for get_pid()

loganharbour committed:
Add tests for start_process()

refs #31755

loganharbour committed:
Run process in the specified directory

refs #31755

loganharbour committed:
Add tests for is_process_running() and remove shell usage

refs #31755

loganharbour committed:
Remove unused argument

loganharbour committed:
Rename SubprocessRunnerBase to SubprocessRunnerInterface

refs #31755

loganharbour committed:
Add moose test for /kill

refs #31755

loganharbour committed:
Rename test based on class rename

loganharbour committed:
Cleanup timeouts and add testing for timeouts

refs #31755

loganharbour committed:
Add an initial poke

refs #31755

loganharbour committed:
Make timeout tests less rigid

refs #31755

loganharbour committed:
Add pyproject.toml for ruff and black

loganharbour committed:
Make ruff happy

moosebuild commented:
All [jobs](https://civet.inl.gov/event/262524/) on b8b877f : canceled by @loganharbour

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3323659/), step WIP Python: black format on 4c34de6 wanted to post the following:

### Python black formatting

Your code requires style changes.

A patch was generated and copied [here](https://mooseframework.inl.gov/docs/PRs/31756/black/black.patch).

You can directly apply the patch by running the following at the top level of your repository:

```
curl -s https://mooseframework.inl.gov/docs/PRs/31756/black/black.patch | git apply -v
```

Alternatively, you can run the following at the top level of your repository:

```
black --config pyproject.toml .
```

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3324390/), step Results summary on ffe1510 wanted to post the following:

## Framework test summary

### Removed tests
None
### Added tests
| Test                                          |   Time (s) |
|-----------------------------------------------|------------|
| `controls/web_server_control.timeout`         |       2.23 |
| `controls/web_server_control.kill`            |       2.21 |
| `controls/web_server_control.initial_timeout` |       2.01 |
### Run time changes
| Test                                                                      |   Base (s) |   Head (s) | +/-     |
|---------------------------------------------------------------------------|------------|------------|---------|
| `outputs/debug.show_execution_order/linear_fv_elem_kernels`               |       0.78 |       1.46 | +88.74% |
| `functions/piecewise_multilinear.oneDb`                                   |       0.81 |       1.43 | +77.18% |
| `meshgenerators/combiner_generator.translations/no_merge_subdomains`      |       0.82 |       1.43 | +73.94% |
| `outputs/intervals.no_intermediate`                                       |       0.92 |       1.57 | +70.48% |
| `controls/web_server_control.errors/set_controllable_vector_non_array`    |       1.25 |       2.12 | +69.47% |
| `materials/derivative_material_interface.new_getproperty_semantics`       |       0.79 |       1.33 | +68.51% |
| `controls/web_server_control.errors/set_controllable_no_exist`            |       1.24 |       2.08 | +67.77% |
| `preconditioners/hmg.hmg`                                                 |       0.96 |       1.6  | +66.39% |
| `materials/material.test`                                                 |       0.88 |       1.46 | +65.58% |
| `variables/fe_hier.group/test_hier_2_3d`                                  |       0.97 |       1.59 | +64.24% |
| `controls/web_server_control.errors/set_controllable_matrix_bad_convert1` |       1.36 |       2.2  | +62.00% |
| `functions/piecewise_constant_from_csv.csv_read/periodic_voronoi`         |       0.85 |       1.36 | +60.50% |
| `controls/web_server_control.errors/postprocessor_no_exist`               |       1.34 |       2.14 | +60.37% |
| `geomsearch/patch_update_strategy.always-grid-sequencing`                 |       1.34 |       2.14 | +59.60% |
| `positions.test/functor_extrema`                                          |       0.79 |       1.26 | +58.86% |
| `functions/solution_function.rot4`                                        |       0.86 |       1.36 | +57.60% |
| `misc/data_file.show_data_params`                                         |       0.9  |       1.41 | +56.88% |
| `controls/web_server_control.errors/set_controllable_matrix_bad_convert2` |       1.4  |       2.19 | +56.57% |
| `controls/web_server_control.errors/set_controllable_bad_convert_json`    |       1.33 |       2.06 | +54.99% |
| `geomsearch/3d_moving_penetration.pl_test2tt`                             |       1.04 |       1.61 | +54.41% |
| `transfers/multiapp_nearest_node_transfer.multiple_target_boundaries`     |       0.97 |       1.49 | +54.19% |
| `misc/deprecation.blessed_coupled_var`                                    |       0.85 |       1.3  | +53.51% |
| `userobjects/shape_element_user_object.diagonal/kernels`                  |       1.32 |       2.02 | +53.10% |
| `meshgenerators/mesh_diagnostics_generator.generate/nonconformality_3d`   |       0.78 |       1.2  | +52.86% |
| `controls/web_server_control.errors/set_controllable_matrix_jagged`       |       1.37 |       2.09 | +52.77% |
| `postprocessors/element_extreme_material_property.ad`                     |       0.83 |       1.26 | +52.58% |
| `meshgenerators/elements_to_tetrahedrons_convertor.sideset_retain/prism6` |       0.84 |       1.29 | +52.24% |
| `meshgenerators/add_meta_data_generator.err_dup_point_scalar_name`        |       1.23 |       1.86 | +51.45% |
| `bcs/periodic.testwedgesys`                                               |       0.89 |       1.35 | +51.22% |
| `kernels/array_kernels.save_in/test_standard_save_in`                     |       0.95 |       1.43 | +51.15% |
| `meshgenerators/cut_mesh_by_plane_generator.test/simple_prism_cut`        |       0.76 |       1.15 | +51.08% |
| `misc/multiple-nl-systems.restart/restart_problem`                        |       0.84 |       1.27 | +51.00% |
| `userobjects/writable_variable.block2`                                    |       0.82 |       1.24 | +50.81% |
| `meshgenerators/xy_delaunay_generator.xydelaunay/nested`                  |       0.81 |       1.22 | +50.63% |
| `geomsearch/penetration_locator.test`                                     |       0.86 |       1.3  | +50.49% |
| `meshgenerators/file_mesh_generator.exodus_iga_nosplines_test`            |       0.8  |       1.2  | +50.22% |
| `meshgenerators/rename_boundary_generator.merge`                          |       0.88 |       1.32 | +50.01% |
## Modules test summary

### Removed tests
None
### Added tests
None
### Run time changes
| Test                                                                                                             |   Base (s) |   Head (s) | +/-     |
|------------------------------------------------------------------------------------------------------------------|------------|------------|---------|
| `stochastic_tools/test:samplers/mcmc.mcmc_sampling/pmcmc_base_bounds`                                            |       3.03 |       5.47 | +80.49% |
| `stochastic_tools/test:samplers/mcmc.mcmc_sampling/pmcmc_base`                                                   |       3.04 |       5.35 | +75.72% |
| `solid_mechanics/test:gravity.ad_gravity_test`                                                                   |       0.84 |       1.41 | +68.78% |
| `stochastic_tools/test:reporters/basic.reporters/ci`                                                             |       0.8  |       1.34 | +67.48% |
| `thermal_hydraulics/test:components/heat_transfer_from_heat_structure_1phase.jac:cylindrical_bottom_side_1phase` |       0.98 |       1.63 | +65.81% |
| `stochastic_tools/test:surrogates/nearest_point.nearest_point/predictors`                                        |       0.8  |       1.28 | +60.80% |
| `navier_stokes/test:finite_element/ins/bcs/advection_bc.check_more_components_than_mesh_dim`                     |       0.89 |       1.41 | +57.79% |
| `ray_tracing/test:outputs/ray_tracing_mesh_output.nemesis_transient`                                             |       0.8  |       1.26 | +57.61% |
| `peridynamics/test:generalized_plane_strain.planestrain_prescribed_OSPD`                                         |       1.09 |       1.71 | +56.32% |
| `subchannel/scripts.index`                                                                                       |       1.42 |       2.19 | +54.84% |
| `porous_flow/test:hysteresis.hys_pc_2_exponential_power`                                                         |       0.86 |       1.32 | +54.18% |
| `stochastic_tools/test:samplers/mcmc.mcmc_sampling/ss_5prop`                                                     |       4.57 |       7.04 | +53.93% |
| `solid_mechanics/test:umat/shear_order.shear_order_umat_moose_z`                                                 |       1.12 |       1.7  | +52.43% |
| `chemical_reactions/test:thermochimica.wrong_species1`                                                           |       0.96 |       1.47 | +52.33% |
| `chemical_reactions/test:parser.equilibrium_action`                                                              |       0.86 |       1.31 | +52.26% |
| `porous_flow/examples/multiapp_fracture_flow/diffusion_multiapp.fracture_app`                                    |       0.97 |       1.48 | +52.11% |
| `scalar_transport/test:ncp-lms.interpolated_lm_ncp_nodal_constraint_enforcement`                                 |       0.93 |       1.41 | +51.69% |
| `peridynamics/test:simple_tests.2D_irregularD_variableH_BPD`                                                     |       0.92 |       1.39 | +51.28% |
| `heat_transfer/test:generate_radiation_patch.generate_radiation_patch_grid_2D_overpart`                          |       0.78 |       1.19 | +51.26% |
| `misc/test:sensor_postprocessor.steady_thermocouple_sensor`                                                      |       0.81 |       1.23 | +51.21% |
| `ray_tracing/test:traceray/hex20.tests/ete`                                                                      |       0.89 |       1.34 | +51.16% |
| `peridynamics/test:generalized_plane_strain.gps_squares_OSPD`                                                    |       1.71 |       2.59 | +51.07% |
| `phase_field/test:conserved_noise.integral_normal_masked`                                                        |       0.85 |       1.29 | +50.79% |
| `optimization/test:functions/parameter_mesh.parameter_mesh/create_dg`                                            |       0.8  |       1.2  | +50.38% |
| `ray_tracing/test:traceray/quad9.tests/side_aq`                                                                  |       0.82 |       1.23 | +50.24% |
| `peridynamics/test:failure_tests.2D_bond_status_convergence_BPD`                                                 |       1.38 |       2.07 | +50.24% |
| `stochastic_tools/test:samplers/mcmc.mcmc_sampling/des_5prop_var`                                                |       4.54 |       6.82 | +50.08% |

loganharbour committed:
Don't bother with getattr

refs #31755

loganharbour committed:
Skip valgrind for tests

refs #31755

loganharbour committed:
Rename methods to simplify

refs #31755

loganharbour committed:
Begin work on rewriting MooseControl documentation

Refs #31755

loganharbour committed:
Build shared_ptr objects with make_shared

refs #31755

loganharbour committed:
Call initial setup on controls

refs #31755

loganharbour committed:
Don't do initial setup here

loganharbour committed:
Rewrite WebServerControl with thread safety and better registration

refs #31755

loganharbour committed:
Remove cerr

refs #31755

loganharbour committed:
Correct error message

refs #31755

loganharbour committed:
Rename action and add documentation

refs #31755

loganharbour committed:
Apply clang-format

loganharbour committed:
Fix options and lock down flags

refs #31755

loganharbour committed:
Try to make compilers happy; rework constructors

refs #31755

loganharbour committed:
Update test method

refs #31755

loganharbour committed:
Remove now un-needed unit test

loganharbour committed:
Continue work on documentation

refs #31755

loganharbour committed:
Remove test comments

refs #31755

loganharbour committed:
Make ruff and black happy

refs #31755

loganharbour committed:
Fix doco

loganharbour committed:
Reorganize and improve comments

refs #31755

loganharbour committed:
Improve includes

loganharbour committed:
Simplify test

loganharbour committed:
Fix comment

zachmprince commented on python/moosecontrol/validation.py:
```diff
@@ -0,0 +1,184 @@
+        return self.data["execute_on_flags"]
+
+
+@staticmethod
```
```suggestion
```

zachmprince commented on python/moosecontrol/validation.py:
```diff
@@ -0,0 +1,184 @@
+    """
+    assert isinstance(ws_response, WebServerControlResponse)
+    if optional is None:
+        optional = []
```
Why not have the argument default just be a `[]` instead of `None`?

zachmprince commented on python/moosecontrol/runners/socketrunner.py:
```diff
@@ -0,0 +1,127 @@
+        # Whether or not we've used the socket (used in cleanup())
+        self._socket_used: bool = False
+
+        super().__init__(**kwargs)
```
```suggestion
        super().__init__(*args, **kwargs)
```

zachmprince commented on python/moosecontrol/runners/utils/timedpoller.py:
```diff
@@ -0,0 +1,191 @@
+    >>>> helper.poll(should_exit)
+    >>>>
+    >>>> # Finalize timing if it succeeds
+    >>>> helper.end()
```
```suggestion
    >>> helper = TimedPoller(poll_time=0.1, timeout_time=10)
    >>> helper.start()
    >>>
    >>> # Develop some criteria on when to exit
    >>> should_exit = lambda: True # fake criteria
    >>>
    >>> # Poll until should_exit is true, which can be done
    >>> # multiple times. If the timeout time is reached,
    >>> # these calls to poll() will raise TimedPollTimeout
    >>> helper.poll(should_exit)
    >>> helper.poll(should_exit)
    >>>
    >>> # Finalize timing if it succeeds
    >>> helper.end()
```

zachmprince commented on python/moosecontrol/runners/interfaces/subprocessrunnerinterface.py:
```diff
@@ -0,0 +1,237 @@
+
+        Must be overridden.
+        """
+        raise NotImplementedError  # pragma: no cover
```
I don't think it is necessary for this to be abstract. You could just return `[]`. Therefore, the class probably doesn't need to be abstract.

loganharbour replied:
It doesn't really make sense to use this class on its own, but I guess

loganharbour replied:
Actually yeah, you always need to specify an override to how to connect to the server in the spanwed process. It should always be set. If someone wants to not set it, they can have it really return `[]` in derived if desired.

zachmprince commented on python/moosecontrol/tests/test_moosecontrol.py:
```diff
@@ -0,0 +1,709 @@
+        Test initialize().
+
+        Should call the underlying runner's initialize() and wait()
+        and query the daat from initialize.
```
```suggestion
        and query the data from initialize.
```

zachmprince commented on python/moosecontrol/tests/test_moosecontrol.py:
```diff
@@ -0,0 +1,709 @@
+                "RealEigenMatrix",
+                "realeigenmatrix",
+            )
+        self.assertEqual(str(e.exception), "Value not convertible to a 1- or 2-D array")
```
```suggestion
        with self.assertRaisesRegex(
            ValueError, "Value not convertible to a 1- or 2-D array"
        ):
            self.run_test_set_controllable(
                value,
                "RealEigenMatrix",
                "realeigenmatrix",
            )
```

zachmprince commented on python/moosecontrol/tests/test_runners_baserunner.py:
```diff
@@ -0,0 +1,500 @@
+
+def check_baserunner_cleanup_live(
+    test: MooseControlTestCase,
+    runner: BaseRunner,
```
This argument isn't used.

zachmprince commented on python/moosecontrol/tests/test_runners_baserunner.py:
```diff
@@ -0,0 +1,500 @@
+            str(e_cm.exception),
+            f"Initialization timed out after "
+            f"{runner._initialize_poller.total_time:.2f} seconds",
+        )
```
```suggestion
        with self.assertRaisesRegex(
            InitializeTimeout,
            "Initialization timed out after "
            f"{runner._initialize_poller.total_time:.2f} seconds",
        ):
            runner.initialize_poll(lambda: False)
```

zachmprince reviewed (changes requested):
It's so beautiful!

loganharbour commented on python/moosecontrol/validation.py:
```diff
@@ -0,0 +1,184 @@
+def check_response_data(
+    ws_response: WebServerControlResponse,
+    expected: list[Tuple[str, Type | Any | Tuple[Type]]],
+    optional: Optional[list[Tuple[str, Type | Any | Tuple[Type]]]] = None,
```
Any is on purpose here. While it negates the other type, the literal value `Any` in the list means allow any type.

loganharbour committed:
Address review

- Make tests a module and use relative imports
- Remove need to add moose pythonpath for each test
- Replace assertRaises with assertRaisesRegex
- Docstring fixes
- Check and try moose exe once up front in conftest.py

refs #31755

loganharbour committed:
Simplify logic

loganharbour committed:
Print out when using and not using moose

refs #31755

loganharbour committed:
Simplify argument

refs #31755

loganharbour commented on python/doc/content/python/moosecontrol/index.md:
```diff
@@ -0,0 +1,158 @@
+# MooseControl
```
I'm not quite sure I want to address this yet. I would still like to do more work on setting up the logger configuration to make it easier. The vebose and quiet is just the basic config
